### ISSUE TITLE
Wait a moment before adding a new song

### ISSUE THREAD

**Issue by jfrankenau** [joshkunz/ashuffle#13]
I am using the Android client [M.A.L.P.](https://gitlab.com/gateship-one/malp) which has the option of replacing the current playlist with an album. I assume that it uses something like `mpc clear; mpc add ALBUM` (which is how I do it in scripts as well) and I can't think of a better way to do this. In conjunction with ashuffle however this results in a random song being added inbetween because for a short moment the playlist is empty and thus ashuffle adds a new random song followed by MALP adding its album.

I'd like to propose that when ashuffle detects an empty playlist to have it wait a few milliseconds and check if the playlist still is empty and only then add a new random song.

**Comment by notEvil** [comment]
+1, `mpc clear` followed by `mpc add` or alike and finally `mpc save` is the intended process to create a playlist so imo ashuffle shouldn't interfere. Is there any good reason why this is still not addressed?

**Comment by notEvil** [comment]
This got me thinking, what if ashuffle could use this as a feature instead of just working around MALP. Consider the following:

When ashuffle receives a `clear` it will wait for a couple ms and check if the playlist is empty. If it is, then do the usual things like add songs and start playing. However when the playlist is not empty, pause operation entirely and wait for the next `clear` followed by empty playlist to resume operation.

Of course, this is entirely different behavior and might conflict with ashuffles current mode of operation. But it would open up a lot of use cases where you don't want ashuffle to add random songs, like when playing a single album on purpose.

**Comment by notEvil** [comment]
Just tried the following stupid implementation and it seems to work :)

```diff
diff --git a/src/ashuffle.cc b/src/ashuffle.cc
index 67415ed..31b5ad3 100644
--- a/src/ashuffle.cc
+++ b/src/ashuffle.cc
@@ -21,6 +21,9 @@
 #include "shuffle.h"
 #include "util.h"
 
+#include <chrono>
+#include <thread>
+
 namespace ashuffle {
 
 namespace {
@@ -149,6 +152,7 @@ void Loop(mpd::MPD *mpd, ShuffleChain *songs, const Options &options,
         TryEnqueue(mpd, songs, options);
     }
 
+    bool suspended = false;
     // Loop forever if test delegates are not set.
     while (test_d.until_f == nullptr || test_d.until_f()) {
         /* wait till the player state changes */
@@ -162,7 +166,15 @@ void Loop(mpd::MPD *mpd, ShuffleChain *songs, const Options &options,
             std::cout << "Picking random songs out of a pool of "
                       << songs->Len() << "." << std::endl;
         } else if (events.Has(MPD_IDLE_QUEUE) || events.Has(MPD_IDLE_PLAYER)) {
-            TryEnqueue(mpd, songs, options);
+            std::unique_ptr<mpd::Status> status = mpd->CurrentStatus();
+            if (status->QueueLength() == 0) {
+                std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+                status = mpd->CurrentStatus();
+                suspended = status->QueueLength() != 0;
+            }
+            if (!suspended) {
+                TryEnqueue(mpd, songs, options);
+            }
         }
     }
 }
```

**Comment by joshkunz** [comment]
I'm a little concerned about any modifications that will make the enqueue slower, even if it's only in the case that the queue is totally cleared. That said, I'm happy to support this if it's hidden by a flag (seems like a great use for a new tweak option). One important thing to keep in mind with the implementation will be testing it, since testing delay can be tricky.

Anyways, I'll get around to adding this soon. It's not that complicated. As always PRs are welcome if you want this done sooner 😄

**Comment by notEvil** [comment]
Off topic, but maybe worth a look: https://gitlab.com/notEvil/mpd-shuffle (I struggled to implement an idea and then switched to Python with quick success)

**Comment by joshkunz** [comment]
@notEvil's change has been merged as `v3.9.0` which has now been released. The [readme has instructions](https://github.com/joshkunz/ashuffle#advanced-options-for-specialized-preferences-with---tweak) for the new `suspend-timeout` tweak.

**Comment by notEvil** [comment]
Thanks a lot, much appreciated!
